src/zonetree/answer.rs

```diff
@@ -0,0 +1,217 @@
+        let qname = question.qname();
+        let qclass = question.qclass();
+        let mut builder = builder.start_answer(message, self.rcode).unwrap();
+
+        match self.content {
+            AnswerContent::Data(ref answer) => {
+                for item in answer.data() {
+                    // TODO: This will panic if too many answers were given,
+                    // rather than give the caller a way to push the rest into
+                    // another message.
+                    builder
+                        .push((qname, qclass, answer.ttl(), item))
+                        .unwrap();
+                }
+            }
+            AnswerContent::Cname(ref cname) => builder
+                .push((qname, qclass, cname.ttl(), cname.data()))
+                .unwrap(),
+            AnswerContent::NoData => {}
+        }
+
+        let mut builder = builder.authority();
+        if let Some(authority) = self.authority.as_ref() {
+            if let Some(soa) = authority.soa.as_ref() {
+                builder
+                    .push((
+                        authority.owner.clone(),
+                        qclass,
+                        soa.ttl(),
+                        soa.data(),
+                    ))
+                    .unwrap();
+            }
+            if let Some(ns) = authority.ns.as_ref() {
+                for item in ns.data() {
+                    builder
+                        .push((
+                            authority.owner.clone(),
+                            qclass,
+                            ns.ttl(),
+                            item,
+                        ))
+                        .unwrap()
+                }
+            }
+            if let Some(ref ds) = authority.ds {
+                for item in ds.data() {
+                    builder
+                        .push((
+                            authority.owner.clone(),
+                            qclass,
+                            ds.ttl(),
+                            item,
+                        ))
+                        .unwrap()
+                }
+            }
+        }
+
+        builder.additional()
+    }
+
+    /// Gets the [`Rcode`] for this answer.
+    pub fn rcode(&self) -> Rcode {
+        self.rcode
+    }
+
+    /// Gets the answer section content for this answer.
+    pub fn content(&self) -> &AnswerContent {
+        &self.content
+    }
+
+    /// Gets the authority section content for this answer.
+    pub fn authority(&self) -> Option<&AnswerAuthority> {
+        self.authority.as_ref()
+    }
+}
+
+//------------ AnswerContent -------------------------------------------------
+
+/// The content of the answer.
+#[derive(Clone)]
+pub enum AnswerContent {
+    Data(SharedRrset),
+    Cname(SharedRr),
+    NoData,
+}
+
+//------------ AnswerAuthority -----------------------------------------------
+
+/// The authority section of a query answer.
+#[derive(Clone)]
+pub struct AnswerAuthority {
```

Comment:
This seems to be a set of things that `Answer::to_message()` knows how to handle, rather than arbitrary authority section records, i.e. it isn't intended to do as you suggest. @partim?